lib/libzfs/libzfs_util.c

```diff
@@ -1940,3 +1941,99 @@ zprop_iter(zprop_func func, void *cb, boolean_t show_all, boolean_t ordered,
 {
 	return (zprop_iter_common(func, cb, show_all, ordered, type));
 }
+
+/*
+ * Utility function to guarantee malloc() success.
+ */
+static void *
+safe_malloc(size_t size)
+{
+	void *data;
+
+	if ((data = calloc(1, size)) == NULL)
+		errno = ENOMEM;
+
+	return (data);
+}
+
+#ifndef ZFS_VERSION_EXTRA
+#define ZFS_VERSION_EXTRA ""
+#endif
+/*
+ * Returns the zfs userland version with ZFS_META_RELEASE and ZFS_META_GITREV
+ * Returned char * must be freed by the caller.
+ * Returns NULL on error with errno set accordingly
+ */
+char *
+zfs_version_userland(void)
+{
+	char *version = safe_malloc(256);
+	if (version == NULL)
+		goto out;
+
+	(void) snprintf(version, 256, "%s%s",
+	    ZFS_META_GITREV, ZFS_VERSION_EXTRA);
+
+out:
+	return (version);
+}
+
+/*
+ * Returns the zfs kernel version read from "/sys/module/zfs/version"
+ * Returned char * must be freed by the caller.
+ * Returns NULL on error with errno set accordingly
+ */
+char *
+zfs_version_kernel(void)
+{
+	char *f_zversion = "/sys/module/zfs/version";
+	int fd_kmod_version;
+	char *version = safe_malloc(256);
+
+	if (version == NULL)
+		return (NULL);
+
+	if ((fd_kmod_version = open(f_zversion, O_RDONLY)) == -1)
+		goto error;
+
+	if (read(fd_kmod_version, version, 256) == -1)
+		goto error;
+
+	if (close(fd_kmod_version) == -1)
+		goto error;
+
+	return (version);
+
+error:
+	free(version);
+	return (NULL);
+}
+
+/*
+ * Prints both zfs userland and kernel versions
+ */
+void
+zfs_version_print(void)
+{
+	char *zver_userland;
+	char *zver_kernel;
```

Comment:
Since this is all in user space, and the version strings are relatively short I don't think it would be unreasonable to put them on the stack.  This would ensure the allocation couldn't fail and remove the need for the `zfs_malloc()` function and a lot of the error handling.

```
int zfs_version_userland(char *version, int len);
int zfs_version_userland(char *version, int len);
```

Reply:
I'm just going to drop the concept of ZFS_VERSION_EXTRA altogether.  I wanted to suffix the version string, but trying to use ZFS_META_RELEASE will wedge in whatever I add into the middle of ZFS_META_GITREV.  At this point, it's just not worth it.